modules/transaction-payment/src/lib.rs

```diff
@@ -813,43 +758,98 @@ where
+			Some(Call::with_fee_currency { currency_id, .. }) => {
+				ensure!(
+					TokenExchangeRate::<T>::contains_key(currency_id),
+					Error::<T>::InvalidToken
+				);
+				Self::swap_from_pool_or_dex(who, custom_fee_amount, *currency_id).map(|_| custom_fee_surplus)
+			}
+			Some(_) | None => Self::native_then_alternative_or_default(who, fee),
+		}
 	}
 
-	/// Iterate order list, break if can swap out enough native asset amount with user's foreign
-	/// asset. make sure trading path is exist in dex, if the trading pair is not exist in dex, even
-	/// though we have setup it in charge fee pool, we can't charge fee with this foreign asset.
-	fn swap_native_asset(who: &T::AccountId, amount: Balance) {
-		let native_currency_id = T::NativeCurrencyId::get();
-		for trading_path in Self::get_trading_path(who) {
-			if trading_path.last() == Some(&native_currency_id) {
-				let supply_currency_id = *trading_path.first().expect("should match a non native asset");
-				if T::MultiCurrency::free_balance(supply_currency_id, who).is_zero() {
-					continue;
+	/// If native is enough, do nothing, return `Ok(0)` means there are none extra surplus fee.
+	/// If native is not enough, try swap from tx fee pool or dex. As user can set his own
+	/// `AlternativeFeeSwapPath`, this will direct swap from dex. Sometimes, user setting of
+	/// `AlternativeFeeSwapPath` may be wrong or dex is not available, or user do not set any
+	/// `AlternativeFeeSwapPath`, then use the `DefaultFeeTokens` to swap from tx fee pool.
+	fn native_then_alternative_or_default(
+		who: &T::AccountId,
+		fee: PalletBalanceOf<T>,
+	) -> Result<Balance, DispatchError> {
+		if let Some(amount) = Self::check_native_is_not_enough(who, fee) {
+			// native asset is not enough
+			let fee_surplus = T::AlternativeFeeSurplus::get().mul_ceil(fee);
+			let fee_amount = fee_surplus.saturating_add(amount);
+
+			// alter native fee swap path, swap from dex: O(1)
+			if let Some(path) = AlternativeFeeSwapPath::<T>::get(who) {
+				if T::DEX::swap_with_specific_path(who, &path, SwapLimit::ExactTarget(Balance::MAX, fee_amount)).is_ok()
+				{
+					return Ok(fee_surplus);
+				}
+			}
+
+			// default fee tokens, swap from tx fee pool: O(1)
+			for supply_currency_id in T::DefaultFeeTokens::get() {
+				if Self::swap_from_pool_or_dex(who, fee_amount, supply_currency_id).is_ok() {
+					return Ok(fee_surplus);
 				}
-				if Self::swap_from_pool_or_dex(who, amount, supply_currency_id).is_ok() {
-					break;
+			}
+
+			// migration of `GlobalFeeSwapPath`. after Dapp using `with_fee_currency`, we can delete this.
+			let global_fee_swap_path = GlobalFeeSwapPath::<T>::iter_values()
+				.map(|v| v.into_inner())
+				.collect::<Vec<_>>();
+			for path in global_fee_swap_path {
+				if let Some(supply_currency_id) = path.first() {
+					if Self::swap_from_pool_or_dex(who, fee_amount, *supply_currency_id).is_ok() {
+						return Ok(fee_surplus);
+					}
 				}
 			}
 		}
+
+		// native asset is enough
+		Ok(0)
```

Comment:
I think this should be in an else block to cover the edge case where native is not enough and none of the other options work. Right now it would return Ok(()) but I think it should return an error in that scenario

Reply:
in the case of native is not enough, even return Ok(0), when withdraw_asset, it'll throw error. for fail-fast, return Error now.